### PR TITLE
Add plugin infrastructure

### PR DESCRIPTION
**Pull request by Yurlungur**
## PR Summary

This PR adds a plugin infrastructure to `singularity-eos` so that at configure time, a user may request additional directories to be incorporated into singularity, which may contain additional equations of state, modifiers, C++ or fortran code, and tests. The `singularity` build system should be robust enough to incorporate these and build them and include them as tests.

I would like feedback on the current implementation if you get the chance. @dholladay00 @mauneyc-LANL . More to come soon.

This is currently a work in progress. Checklist:
- [x] Enable plugins to register header, source files and tests
- [x] Example plugin builds, with tests.
- [x] Enable plugins to specify the EOS variant to use
- [x] Thread machinery through spack
- [x] Check edge cases

## PR Checklist

- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Format your changes by using the `make format` command after configuring with `cmake`.
- [x] Document any new features, update documentation for changes made.
- [x] Make sure the copyright notice on any files you modified is up to date.
- [x] After creating a pull request, note it in the CHANGELOG.md file.
- [ ] LANL employees: make sure tests pass both on the github CI and on the Darwin CI

If preparing for a new release, in addition please check the following:
- [ ] Update the version in cmake.
- [ ] Move the changes in the CHANGELOG.md file under a new header for the new release, and reset the categories.

### CODE REVIEW

**Comment by Yurlungur** [comment]
@dholladay00 @mauneyc-LANL @jhp-lanl this is ready for review. The only missing piece now is the spack variant, which I will look at soon.

**Comment by Yurlungur** [comment]
@gopsub this machinery will be used for the eos-extras stuff we discussed elsewhere. I will ping you offline to explain.

**Comment by Yurlungur** [comment]
@rbberger I have added the plugin machinery to spack. Could you take a look and let me know what you think?

File: spack-repo/packages/singularity-eos/package.py
Comment (rbberger) [inline spack-repo/packages/singularity-eos/package.py:203]: Who sets the environment variables and when? Where does the plugins source code live? Instead of a Boolean variant, I would also imagine `plugins` to be something similar to `patches`.

```bash
spack install singularity-eos plugins=dir1:dir2:dir3 variant=some_path/to/another/file
```

I'm not sold on the idea to have this generalized in the public spackage. For the particular use case in mind, what's the downside of just having a inherited Spack package for singularity-eos and manipulating the CMake build and source directory for a particular plugin combo?

As in, Project XYZ has its own `singularity-eos` spackage that inherits from the upstream one, but grabs more source files, puts it in the right place and uses the CMake machinery to make it show up in the Singularity-EOS build.

Comment (Yurlungur) [reply spack-repo/packages/singularity-eos/package.py:203]: Do variants accept arbitrary strings? The reason I went with the environment variable path is that it wasn't clear from the spack documentation that this is possible. It is possible?

> I'm not sold on the idea to have this generalized in the public spackage. For the particular use case in mind, what's the downside of just having a inherited Spack package for singularity-eos and manipulating the CMake build and source directory for a particular plugin combo?

This is fine with me. The advantage of doing it here is that we only have to maintain one spackage. But maybe that's basically the case with the inheritence version as well? Since presumably the inherited package doesn't contain much?

File: cmake/plugins.cmake
Comment (Yurlungur) [reply cmake/plugins.cmake:63]: :+1: Done

File: spack-repo/packages/singularity-eos/package.py
Comment (Yurlungur) [reply spack-repo/packages/singularity-eos/package.py]: Ok I will go the spackage inheritence path for now.

**Comment by Yurlungur** [comment]
@rbberger @dholladay00 @mauneyc-LANL @junghans  please re-review. I believe this is ready. As Richard suggests, I will attempt (as a first pass) to do plugins with spackage inheritence, rather than adding it to the spackage directly. However, I would like feedback on this approach. Is it better to add plugins as a variant? Or better to do spackage via inheritence?

File: CMakeLists.txt
Comment (mauneyc-LANL) [inline CMakeLists.txt:423]: Shouldn't we have an install interface here? I think it'd just be the same as the `$<INSTALL_INTERFACE:include>`, but there's other scaffolding to manage (`INSTALL()` and `EXPORT()`)

**Review by mauneyc-LANL — COMMENTED** [review]
I left some comments, but I need to give it a test drive on my side. We also need to being thinking about how spack integrates into this

File: CMakeLists.txt
Comment (Yurlungur) [reply CMakeLists.txt:397]: Oh cool. I will change this to integrate with the standard cmake version.

File: CMakeLists.txt
Comment (Yurlungur) [reply CMakeLists.txt:423]: I just hooked it into your current install machinery. If you look above, I add headers to the install headers list object. And for the source files, they all get linked into a single library anyway. That additional `foreach` that you comment on here is just to make sure in-tree/submodule mode builds work.

File: CMakeLists.txt
Comment (Yurlungur) [reply CMakeLists.txt]: Done

**Comment by Yurlungur** [comment]
> I left some comments, but I need to give it a test drive on my side.

Cool please let me know at your earliest convenience. There is starting to be a need for the extras repo, so I want this plugin infrastructure merged as quickly as possible.

>  We also need to being thinking about how spack integrates into this

Yes---did you see the discussion with @rbberger above? We should discuss at the standup meeting this week. It seems that there are some differences of opinion about the right path forward.

File: spack-repo/packages/singularity-eos/package.py
Comment (rbberger) [reply spack-repo/packages/singularity-eos/package.py]: I'm not aware of a limitation of variants, they can be arbitrary strings. That does not mean they should be arbitrary complex strings. From my point of view, the main issue here is that spack covers runtime plugin systems (e.g. python/pypi), not compile time plugin systems.

Comment (Yurlungur) [reply spack-repo/packages/singularity-eos/package.py]: :+1: Either way is fine with me, so long as we have something that is useful broadly.

**Review by junghans — APPROVED** [review]

**Comment by mauneyc-LANL** [comment]
Pushed a small change for install to work

**Comment by Yurlungur** [comment]
> Pushed a small change for install to work

Thanks for the fix!

**Comment by Yurlungur** [comment]
Got verbal approval from @mauneyc-LANL so I'm going to click the button now. @rbberger I think we can start on the spackage at this point.

**Comment by rbberger** [comment]
Ok. On it.

**Comment by Yurlungur** [comment]
correction. clicking button after tests pass

File: doc/sphinx/src/customization
Comment (jhp-lanl) [inline doc/sphinx/src/customization:22]: Do we want to add anything here about the helper functions @dholladay00 created for assembling complex type lists? I'd be fine with a small blurb to just go look at `eos.hpp` for examples.

File: doc/sphinx/src/customization
Comment (jhp-lanl) [inline doc/sphinx/src/customization:80]: I really like this

**Review by jhp-lanl — COMMENTED** [review]
Sorry for the feedback after merge. My comments only concern documentation and can be wrapped into an issue for later.

Other than my comment on creating the type list, we should change the instructions for adding a new EOS that _specifically_ reference `eos.hpp` which has been renamed to `default_variant.hpp`. This should be a pretty easy grep and replace in the documentation.

**Pull request merged**
